staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/ratcheting.go

```diff
@@ -0,0 +1,336 @@
+		r.comparisonResult = &res
+	}()
+
+	if r.value == nil && r.thunk.oldValue == nil {
+		return true
+	} else if r.value == nil || r.thunk.oldValue == nil {
+		return false
+	}
+
+	oldAsArray, oldIsArray := r.thunk.oldValue.([]interface{})
+	newAsArray, newIsArray := r.value.([]interface{})
+
+	if oldIsArray != newIsArray {
+		return false
+	} else if oldIsArray {
+		if len(oldAsArray) != len(newAsArray) {
+			return false
+		}
+
+		// Correctly considers map-type lists due to fact that index here
+		// is only used for numbering. The correlation is stored in the
+		// childInvocation itself
+		for i := range newAsArray {
+			// Query for child
+			childInvocation, ok := r.children[i]
+			if !ok {
+				return false
+			} else if !childInvocation.IsNoopLazy() {
+				return false
+			}
+		}
+
+		return true
+	}
+
+	oldAsMap, oldIsMap := r.thunk.oldValue.(map[string]interface{})
+	newAsMap, newIsMap := r.value.(map[string]interface{})
+
+	if oldIsMap != newIsMap {
+		return false
+	} else if oldIsMap {
+		if len(oldAsMap) != len(newAsMap) {
+			return false
+		}
+
+		for k := range oldAsMap {
+			// Query for child
+			childInvocation, ok := r.children[k]
+			if !ok {
+				// Either key in old not in new, or validator short
+				// circuited
+				_, newAlsoHasKey := newAsMap[k]
+				if newAlsoHasKey {
+					// Child invocation is missing due to short circuit
+					return reflect.DeepEqual(r.thunk.oldValue, r.value)
+				}
+
+				// Child invocation is missing due to key change
+				return false
+			} else if !childInvocation.IsNoopLazy() {
+				return false
+			}
+		}
+
+		return true
+	}
+
+	return reflect.DeepEqual(r.thunk.oldValue, r.value)
+}
+
+func newRatchetingValueValidator(oldValue interface{}, args schemaArgs, parent *ratchetingValueValidator, keyInParent interface{}) *ratchetingValueValidator {
+	return &ratchetingValueValidator{
+		oldValue:    oldValue,
+		schemaArgs:  args,
+		parent:      parent,
+		keyInParent: keyInParent,
+	}
+}
+
+func (r *ratchetingValueValidator) validateOption(value interface{}) validate.Option {
+	return func(svo *validate.SchemaValidatorOptions) {
+		svo.NewValidatorForField = func(field string, schema *spec.Schema, rootSchema interface{}, root string, formats strfmt.Registry, opts ...validate.Option) validate.ValueValidator {
+			return r.SubPropertyValidator(value, field, schema, rootSchema, root, formats, opts...)
+		}
+		svo.NewValidatorForIndex = func(index int, schema *spec.Schema, rootSchema interface{}, root string, formats strfmt.Registry, opts ...validate.Option) validate.ValueValidator {
+			return r.SubIndexValidator(value, index, schema, rootSchema, root, formats, opts...)
+		}
+	}
+}
+
+func (r *ratchetingValueValidator) Applies(value interface{}, kind reflect.Kind) bool {
+	return true
+}
+
+func (s *ratchetingValueValidator) SetPath(path string) {
+	s.schemaArgs.Path = path
+}
+
+// Validate validates the value.
+func (r *ratchetingValueValidator) Validate(value interface{}) *validate.Result {
```

Comment:
the side-effect on `r` are surprising. This function needs a go doc describing the general idea of how `r` is used and changed. As far as I see some kind of parallel (double-linked) tree of options is built. Why is that necessary?